zebedee-cms/src/main/java/com/github/onsdigital/zebedee/model/publishing/legacycacheapi/LegacyCacheApiPayloadBuilder.java

```diff
@@ -0,0 +1,205 @@
+                            if (isValid(legacyCacheApiPayload)) {
+                                payloadsByPath.put(legacyCacheApiPayload.uriToUpdate, legacyCacheApiPayload);
+                            }
+                        });
+            } catch (IOException e) {
+                error().data("collectionId", collectionId)
+                        .logException(e, "failed initialising PublishNotification when reading collection reviewedUris");
+            }
+
+            return new LegacyCacheApiPayloadBuilder(this.payloadsByPath.values());
+        }
+
+        private void addPayloadForBulletinLatest(String uri, String collectionId, Date clearCacheDate) {
+            if (uri != null && uri.contains(BULLETIN_SEGMENT) && !uri.endsWith("/latest")) {
+                String updatedURI = URIUtils.getNSegmentsAfterSegmentInput(uri, BULLETIN_SEGMENT, 1) + "/latest";
+                LegacyCacheApiPayload legacyCacheApiPayloadLatest = new LegacyCacheApiPayload(collectionId, updatedURI, clearCacheDate);
+                payloadsByPath.put(legacyCacheApiPayloadLatest.uriToUpdate, legacyCacheApiPayloadLatest);
+            }
+        }
+
+        private boolean isValid(LegacyCacheApiPayload payload) {
+            if (StringUtils.isEmpty(payload.uriToUpdate)) {
+                info().data("collectionId", payload.collectionId)
+                        .log("invalid payload: URI was empty or null");
+                return false;
+            }
+            return true;
+        }
+
+        public static String getCanonicalPagePath(String incomingUrl, String  collectionId) {
+            try {
+                if (StringUtils.isEmpty(incomingUrl)) {
+                    return incomingUrl;
+                }
+
+                String updatedUri = incomingUrl.trim();
+                if (StringUtils.isEmpty(updatedUri)) {
+                    return "";
+                }
+
+                // Order matters:
+                // 1. If starts with ons return original
+                if (updatedUri.startsWith(ONS_PREFIX)) {
+                    return updatedUri;
+                }
+
+                // 2. If starts with /visualisation/ then /visualisation/xxx/xxxx/ ---> /visualisation/xxx (keep 1 segment) - return here
+                if (updatedUri.startsWith(VISUALISATIONS_SEGMENT)) {
+                    return URIUtils.getNSegmentsAfterSegmentInput(updatedUri, VISUALISATIONS_SEGMENT, 1);
+                }
+
+                // 3. If resource then e.g. /embed?uri='/something' ---> /something:
+                //      /chartconfig
+                //      /chartimage
+                //      /embed
+                //      /chart
+                //      /resource
+                //      /generator
+                //      /file
+                //      /export
+                for (String keyword : RESOURCE_TYPE_SEGMENTS) {
+                    if (updatedUri.startsWith(keyword)) {
```

Comment:
You should probably check that it starts with that keyword + `?`. That way you ensure that the path is exactly what you'd expect (for instance, your current logic also matches `files`, which is wrong), and that it has query params.

Reply:
the ? is already in RESOURCE_TYPE_SEGMENTS